### PR TITLE
Compile MKLDNN in Mac.

### PR DESCRIPTION
**Pull request by zheng-da**
## Description ##
This is to make MKLDNN compiled in Mac correctly.

### Essentials ###
- [ ] Passed code style checking (`make lint`)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- [ ] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### CODE REVIEW

File: prepare_mkldnn.sh [hunk 1]
```diff
@@ -84,7 +84,13 @@ if [ ! -f "$MKLDNN_INSTALLDIR/lib/libmkldnn.so" ]; then
...
 	g++ --version >&2
     cmake $MKLDNN_ROOTDIR -DCMAKE_INSTALL_PREFIX=$MKLDNN_INSTALLDIR -B$MKLDNN_BUILDDIR -DARCH_OPT_FLAGS="-mtune=generic" >&2
-    make -C $MKLDNN_BUILDDIR -j$(cat /proc/cpuinfo | grep processor | wc -l) VERBOSE=1 >&2
+    if [ $OSTYPE == "linux-gnu" ]; then
+      make -C $MKLDNN_BUILDDIR -j$(cat /proc/cpuinfo | grep processor | wc -l) VERBOSE=1 >&2
+    elif [ $OSTYPE == "darwin16" ]; then
+      make -C $MKLDNN_BUILDDIR -j$(sysctl -n hw.ncpu) VERBOSE=1 >&2
+    else
+      make -C $MKLDNN_BUILDDIR VERBOSE=1 >&2
+    fi
```
Comment (szha) [inline prepare_mkldnn.sh]: Maybe refactor the processor discovery to reduce duplicated make calls? I used this snippet for discovering number of processes elsewhere.
```bash
NUM_PROC=1
if [[ ! -z $(command -v nproc) ]]; then
    NUM_PROC=$(nproc)
elif [[ ! -z $(command -v sysctl) ]]; then
    NUM_PROC=$(sysctl -n hw.ncpu)
else
    >&2 echo "Can't discover number of cores."
fi
```

File: prepare_mkldnn.sh [hunk 2]
```diff
@@ -84,7 +84,13 @@ if [ ! -f "$MKLDNN_INSTALLDIR/lib/libmkldnn.so" ]; then
     cd $MXNET_ROOTDIR
 	g++ --version >&2
     cmake $MKLDNN_ROOTDIR -DCMAKE_INSTALL_PREFIX=$MKLDNN_INSTALLDIR -B$MKLDNN_BUILDDIR -DARCH_OPT_FLAGS="-mtune=generic" >&2
-    make -C $MKLDNN_BUILDDIR -j$(cat /proc/cpuinfo | grep processor | wc -l) VERBOSE=1 >&2
+    if [ $OSTYPE == "linux-gnu" ]; then
+      make -C $MKLDNN_BUILDDIR -j$(cat /proc/cpuinfo | grep processor | wc -l) VERBOSE=1 >&2
```
Comment (lebeg) [inline prepare_mkldnn.sh]: I think -j$(nproc) will work on most platforms

File: prepare_mkldnn.sh [hunk 3]
```diff
@@ -84,7 +84,13 @@ if [ ! -f "$MKLDNN_INSTALLDIR/lib/libmkldnn.so" ]; then
...
+    elif [ $OSTYPE == "darwin16" ]; then
+      make -C $MKLDNN_BUILDDIR -j$(sysctl -n hw.ncpu) VERBOSE=1 >&2
+    else
+      make -C $MKLDNN_BUILDDIR VERBOSE=1 >&2
```
Comment (lebeg) [inline prepare_mkldnn.sh]: Which OS is the else branch? Why multiple cores can not be used?

File: prepare_mkldnn.sh (hunk 2 above)
Comment (zheng-da) [reply prepare_mkldnn.sh]: `make -j$(nproc)` doesn't seem to work on Mac

File: prepare_mkldnn.sh (hunk 3 above)
Comment (zheng-da) [reply prepare_mkldnn.sh]: I only listed two popular OS. I don't know what other OS people use and what mechanism is provided for getting the number of cores automatically. We need to handle the default case.

**Comment by szha** [comment]
Given that our CI doesn't have mac yet, please verify the new logic locally.

**Comment by zheng-da** [comment]
I tested it. It works fine.

**Comment by szha** [comment]
Does the current logic require re-building mkldnn at every `make`? Is caching the build artifact possible?

**Comment by zheng-da** [comment]
Yes, it's weird. We don't need to rebuild mkldnn every time in Linux.

**Comment by szha** [comment]
This needs to be fixed.

**Comment by zheng-da** [comment]
it should work now.

**Comment by pengzhao-intel** [comment]
FYI, as @KellenSunderland @marcoabreu 's suggestion, we added Clang + MKLDNN instance in CI process #9918 (#9828) to track the OSX building issues.

**Comment by szha** [comment]
clang build is different from building on mac. For example, clang build won't catch the processor detection problem.

**Pull request merged**